### DISCUSSION TITLE
`tbl-subcap` with LaTeX tables

### DISCUSSION THREAD

**Discussion by vincentarelbundock** [quarto-dev/quarto-cli#10324] [Q&A]
### Description

Hi,

Thanks for all your work on Quarto and on this forum!

When a package produces markdown tables, the `tbl-subcap` functionality works as expected. However, when the package produces a LaTeX table, there is no subcaption.

Is it possible for Quarto to support subcaptions in LaTeX tables as well?

I'd love to get some clarification on this.

1. Are subcaptions only supported for markdown tables?
2. Are there plans to extend support for LaTeX tables as well?

Thanks!

````
---
format: pdf
---

```{r}
#| label: tbl-example
#| tbl-cap: "Example"
#| tbl-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout-ncol: 2

library(tinytable)
tt(head(cars)) |> save_tt("markdown") |> knitr::asis_output()
tt(head(pressure)) |> save_tt("markdown") |> knitr::asis_output()
```
````

This doesn't work:

```r 
library(tinytable)
tt(head(cars)) |> save_tt("latex") |> knitr::asis_output()
tt(head(pressure)) |> save_tt("latex") |> knitr::asis_output()
```

**Comment by mcanouil** [comment]
It was closed because an issue has been opened as mentioned in the same thread.

- https://github.com/quarto-dev/quarto-cli/discussions/10088#discussioncomment-9843375

Namely:

- #10090

I am not sure to understand the point of this new thread.

**Reply by vincentarelbundock** [reply]
lol, ok. I think I confused two threads. Sorry about that. My question is unrelated to the math issue. I edited it to remove mention of the other thread.

The question: Can we use `tbl-subcap` with raw LaTeX tables, or does Quarto absolutely need to receive a markdown input, and to convert it itself?

**Reply by cscheid** [reply]
Can you fix your example in the post (look at the `options(...)` line) so we can test it ourselves?

**Reply by vincentarelbundock** [reply]
done

**Reply by cscheid** [reply]
Your LaTeX code is emitting `\begin{table}` pairs. We previously discussed this, but the short of it is that in order for raw latex to be consumed correctly by Quarto, it needs to start from `tabular`, `tblr`, or something else. If you add the `table` environment, then Quarto can't support it.

**Reply by cscheid** [reply]
This is the `keep-md: true` output I'm seeing:

````
---
format: pdf
keep-md: true
keep-tex: true
---


::: {#tbl-example .cell layout-ncol="2" tbl-cap='Example' tbl-subcap='["Cars","Pressure"]'}

```{.r .cell-code}
library(tinytable)
```

::: {.cell-output .cell-output-stderr}

```
Warning: package 'tinytable' was built under R version 4.3.3
```


:::

```{.r .cell-code}
tt(head(cars)) |> save_tt("latex") |> knitr::asis_output()
```

::: {.cell-output-display}
\begin{table}
\centering
\begin{tblr}[         %% tabularray outer open
]                     %% tabularray outer close
{                     %% tabularray inner open
colspec={Q[]Q[]},
}                     %% tabularray inner close
\toprule
speed & dist \\ \midrule %% TinyTableHeader
4 &  2 \\
4 & 10 \\
7 &  4 \\
7 & 22 \\
8 & 16 \\
9 & 10 \\
\bottomrule
\end{tblr}
\end{table}
:::

```{.r .cell-code}
tt(head(pressure)) |> save_tt("latex") |> knitr::asis_output()
```

::: {.cell-output-display}
\begin{table}
\centering
\begin{tblr}[         %% tabularray outer open
]                     %% tabularray outer close
{                     %% tabularray inner open
colspec={Q[]Q[]},
}                     %% tabularray inner close
\toprule
temperature & pressure \\ \midrule %% TinyTableHeader
0 & 0.0002 \\
20 & 0.0012 \\
40 & 0.0060 \\
60 & 0.0300 \\
80 & 0.0900 \\
100 & 0.2700 \\
\bottomrule
\end{tblr}
\end{table}
:::
:::
````

In addition to the `\begin{table}...\end{table}` pair, the code isn't wrapped in a raw block element. For Quarto to have the best chance to process your input correctly, you should try to use a single block if at all possible, like this:

````
```{=latex}
\begin{tblr}
...
\end{tblr}
```
````

Note the absence of the `table` environment, and the fact that the LaTeX emitted is wrapped in a raw block.

**Reply by vincentarelbundock** [reply]
I don't think that's the only problem it.

Here's another example, this time using the development version of `tinytable`. I use the `theme_tt()` to remove `\begin{table}` and return a raw `tabular` environment. Yet, it still doesn't work with the subcaption.

First, install the dev version:

```r
remotes::install_github("vincentarelbundock/tinytable")
```

If you look at the saved .tex file, you'll see that this is a raw `tabular` environment:

````r
---
format: 
    pdf:
        keep-tex: true
---

Here's another example, this time using the development version of `tinytable`. I use the `theme_tt()` to remove `\begin{table}` and return a raw `tabular` environment. Yet, it still doesn't work with the subcaption.

```{r}
#| label: tbl-example
#| tbl-cap: "Example"
#| tbl-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout-ncol: 2

library(tinytable)
tt(head(cars)) |> theme_tt("tabular") |> 
  save_tt("latex") |> knitr::asis_output()
tt(head(pressure)) |> theme_tt("tabular")  |> 
  save_tt("latex") |> knitr::asis_output()
```
````

**Reply by vincentarelbundock** [reply]
FWIW, here's the `.md` output of that last example:

````
---
format: pdf
keep-md: true
---

::: {#tbl-example .cell layout-ncol="2" tbl-cap='Example' tbl-subcap='["Cars","Pressure"]'}

```{.r .cell-code}
library(tinytable)
tt(head(cars)) |> theme_tt("tabular") |> 
  save_tt("latex") |> knitr::asis_output()
```

::: {.cell-output-display}
\begin{tabular}{ll}
\hline
speed & dist \\ \hline
4 &  2 \\
4 & 10 \\
7 &  4 \\
7 & 22 \\
8 & 16 \\
9 & 10 \\
\hline
\end{tabular}
:::

```{.r .cell-code}
tt(head(pressure)) |> theme_tt("tabular")  |> 
  save_tt("latex") |> knitr::asis_output()
```

::: {.cell-output-display}
\begin{tabular}{ll}
\hline
temperature & pressure \\ \hline
0 & 0.0002 \\
20 & 0.0012 \\
40 & 0.0060 \\
60 & 0.0300 \\
80 & 0.0900 \\
100 & 0.2700 \\
\hline
\end{tabular}
:::
:::
````

**Reply by cscheid** [reply]
Huh. That's not what I'm seeing with 

```
> install.packages("tinytable")
trying URL 'https://cran.rstudio.com/bin/macosx/big-sur-arm64/contrib/4.3/tinytable_0.3.0.tgz'
Content type 'application/x-gzip' length 513264 bytes (501 KB)
==================================================
downloaded 501 KB


The downloaded binary packages are in
	/var/folders/nm/m64n9_z9307305n0xtzpp54m0000gn/T//Rtmpn7QaV4/downloaded_packages
```

Should I be getting your library elsewhere?

**Reply by vincentarelbundock** [reply]
Yes, as I wrote in my previous post, this requires the development version of tinytable from Github (See `remotes` call above.)

My example uses a function I specifically implemented to deal with a similar issue we had previously discussed.

**Reply by cscheid** [reply]
I can repro the problem, and I think this should work, and I don't know why it doesn't (aka it's a Quarto bug).

I'll note that I still think it'd be safer if you used `` ```{=latex} ... `` (because it guarantees a single LaTeX rawblock as the payload of the cell output), but I tried adding that to your repro document and it doesn't solve the problem.

**Reply by cscheid** [reply]
> Yes, as I wrote in my previous post

I apologize. We go through a lot of these posts, and the folding comments on GitHub tripped me up, and I missed that. In any case, I can repro the bug. I'll open an issue.

**Reply by cscheid** [reply]
Tracking here: https://github.com/quarto-dev/quarto-cli/issues/10328

**Reply by vincentarelbundock** [reply]
> I can repro the problem, and I think this should work, and I don't know why it doesn't (aka it's a Quarto bug).

Thanks for working with me on this. I really appreciate your time!

> I'll note that I still think it'd be safer if you used `` ```{=latex} ... `` (because it guarantees a single LaTeX rawblock as the payload of the cell output), but I tried adding that to your repro document and it doesn't solve the problem.

Oh yes, normally the user will *not* call `save_tt() |> knitr::asis_output()`. In those cases, `tinytable` will automatically wrap things up in `=latex`. I just used that convoluted syntax to show exact parallelism between the markdown and LaTeX examples.
